packages/core/integration-tests/test/cache.js

```diff
@@ -6154,6 +6159,84 @@ describe('cache', function () {
       assert.equal(res, 4);
     });
 
+    it('should write bundle graph to cache on bundling error', async function () {
+      let overlayFSPackageManager = new NodePackageManager(
+        overlayFS,
+        __dirname,
+      );
+      let entries = 'source/index.js';
+      let options = {
+        mode: 'production',
+        defaultTargetOptions: {
+          shouldScopeHoist: false,
+        },
+        packageManager: overlayFSPackageManager,
+        shouldDisableCache: false,
+        inputFS: overlayFS,
+        cacheDir: path.join(__dirname, '.parcel-cache'),
+      };
+
+      await fsFixture(overlayFS)`
+      source
+        foo.js:
+
+          export default 2;
+        index.js:
+          import('./foo');
+
+          export default 1;
+        .parcelrc:
+          {
+            "extends": "@parcel/config-default",
+            "bundler": "./test-bundler.js"
+          }
+        test-bundler.js:
+          import {Bundler} from '@parcel/plugin'
+          import DefaultBundler from '@parcel/bundler-default'
+
+          const CONFIG = Symbol.for('parcel-plugin-config');
+
+          export default new Bundler({
+            loadConfig({config, options}) {
+              return DefaultBundler[CONFIG].loadConfig({config, options});
+            },
+          
+            bundle({bundleGraph, config}) {
+              DefaultBundler[CONFIG].bundle({bundleGraph, config});
+            },
+            optimize() {throw new Error("Intentionally throw error")},
+          });
+        yarn.lock:`;
+
+      let errorThrown = true;
```

Comment:
Instead of a a try/catch and assertion flow, you could use something like `await assert.rejects(() => bundle(entries, options))`